### PR TITLE
Thread safe loader 2

### PR DESCRIPTION
**Pull request by mattp-**
### What does this PR do?
this is a version of #45782 with latest develop merged in. I've not made any changes other than resolving merge conflicts; the original jenkins results have aged out so I'm creating this to re-run. If I figure out a solution I'll send PR to @smarsching and close this.

### Previous Behavior
Remove this section if not relevant

### New Behavior
Remove this section if not relevant

### Tests written?

Yes/No

### Commits signed with GPG?

Yes/No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by damon-atkins** [comment]
Go Go Jenkins!

**Review by DmitryKuzmenko — APPROVED** [review]
LGTM

File: salt/utils/msgpack.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    # Fall back to msgpack_pure
+    import msgpack_pure as msgpack  # pylint: disable=import-error
+
+# Import Salt libs
+import salt.utils.thread_local_proxy
+
+
+def pack(o, stream, **kwargs):
+    '''
+    .. versionadded:: Oxygen
```
Comment (gtmanfred) [inline salt/utils/msgpack.py]: Fluorine

File: salt/utils/msgpack.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    msgpack module using the _msgpack_module argument.
+    '''
+    msgpack_module = kwargs.pop('_msgpack_module', msgpack)
+    o = salt.utils.thread_local_proxy.ThreadLocalProxy.unproxy_recursive(o)
+    return msgpack_module.pack(o, stream, **kwargs)
+
+
+def packb(o, **kwargs):
+    '''
+    .. versionadded:: Oxygen
```
Comment (gtmanfred) [inline salt/utils/msgpack.py]: Fluorine

File: salt/utils/msgpack.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    msgpack module using the _msgpack_module argument.
+    '''
+    msgpack_module = kwargs.pop('_msgpack_module', msgpack)
+    o = salt.utils.thread_local_proxy.ThreadLocalProxy.unproxy_recursive(o)
+    return msgpack_module.packb(o, **kwargs)
+
+
+def unpack(stream, **kwargs):
+    '''
+    .. versionadded:: Oxygen
```
Comment (gtmanfred) [inline salt/utils/msgpack.py]: Fluorine

File: salt/utils/msgpack.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    msgpack_pure, if the msgpack is not available. You can pass an alternate
+    msgpack module using the _msgpack_module argument.
+    '''
+    msgpack_module = kwargs.pop('_msgpack_module', msgpack)
+    return msgpack_module.unpack(stream, **kwargs)
+
+
+def unpackb(packed, **kwargs):
+    '''
+    .. versionadded:: Oxygen
```
Comment (gtmanfred) [inline salt/utils/msgpack.py]: Fluorine

File: salt/utils/thread_local_proxy.py [hunk 5]
```diff
@@ -0,0 +1,738 @@
+# -*- coding: utf-8 -*-
+'''
+Proxy object that can reference different values depending on the current
+thread of execution.
+
+..versionadded:: Oxygen
```
Comment (gtmanfred) [inline salt/utils/thread_local_proxy.py]: Fluorine

**Review by gtmanfred — CHANGES_REQUESTED** [review]
LGTM, needs the versionadded references updated to Fluorine.

[blind-review](https://github.com/zombie/blind-reviews "Review performed in blind mode")

**Comment by gtmanfred** [comment]
Also, looks like there are some pylint warnings in salt/state.py https://jenkinsci.saltstack.com/job/pr-lint/job/PR-48598/1/warnings52Result/

File: salt/transport/zeromq.py [hunk 6]
```diff
@@ -433,7 +433,14 @@ def __del__(self):
     def connect(self):
         if not self.auth.authenticated:
             yield self.auth.authenticate()
-        self.publish_port = self.auth.creds['publish_port']
+
```
Comment (cachedout) [inline salt/transport/zeromq.py]: I'm confused about the relationship between loader thread-safety and these proposed changes to the transports.

**Comment by mattp-** [comment]
Sorry, that accidentally snuck in. It's from my other PR that I put on
develop, not a feature branch. You can ignore.

**Review by thatch45 — APPROVED** [review]

**Comment by cachedout** [comment]
@isbm Does anybody on @saltstack/team-suse want to review this? We're getting close to merging this in.

**Comment by cachedout** [comment]
@mattp- Can you rebase against the head of develop so we have only the related changes? That should clear out those transport changes so we know exactly what we're dealing with here.

**Comment by mattp-** [comment]
@cachedout yep will do shortly. to be clear I haven't made any changes yet; though I did notice one incompatibility between how data.serializers.msgpack and data.utils.msgpack is handled that I need to look at more.

**Comment by mattp-** [comment]
unassociated commit removed

**Comment by cachedout** [comment]
@mattp- Cool. We also need the issues raised by @gtmanfred resolved before we can move forward, please.

**Review by isbm — APPROVED** [review]
LGTM

**Comment by cachedout** [comment]
I spent some time looking at this today and I'm honestly not sure what's happening. It's failing in different ways on different platforms and, oddly enough, it seems to work just fine on my development machine.

**Review by gtmanfred — APPROVED** [review]

**Comment by mattp-** [comment]
closing in favor of #50648

**Pull request closed without merging**